src/AutoMapper/IFeatureCollectionBase.cs

```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace AutoMapper
+{
+    public interface IFeatureCollectionBase<TValue> : IEnumerable<KeyValuePair<Type, TValue>>
+    {
+        /// <summary>
+        /// Gets or sets the feature.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <returns>System.Object.</returns>
+        TValue this[Type key] { get; set; }
+
+        /// <summary>
+        /// Gets the feature of type <typeparamref name="TFeature"/>.
+        /// </summary>
+        /// <typeparam name="TFeature">The type of the feature.</typeparam>
+        /// <returns>The feature or null if feature not exists.</returns>
+        TFeature Get<TFeature>() where TFeature : TValue;
+
+        /// <summary>
+        /// Sets the feature for type <typeparamref name="TFeature"/>.
+        /// </summary>
+        /// <typeparam name="TFeature">The type of the t feature.</typeparam>
+        /// <param name="feature">The feature. To remove a feature, pass null value.</param>
+        void Set<TFeature>(TFeature feature) where TFeature : TValue;
```

Comment:
A collection with Set seems weird. Maybe an Add that disallows duplicates, like Dictionary?

Reply:
If we use `Add` and throw we also need to have an `Remove`-method to be able to replace an already added item. Naming is from the asp.net core FeatureCollection.

Otherwise the `AddOrUpdate` is a common pattern for collections.

Reply:
So this is emulating https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.features.featurecollection?view=aspnetcore-2.2 ?

Reply:
I've never used that, but why would you need to call Set twice for the same extension? It seems simpler to just throw.

Reply:
Yes, it is the item I checked during build, I ignored the `Revision` property due to the usage in .net core where the collection have a "base collection" and then you can replace/add items on that one that not will reflect in the "base collection".

Reply:
In [xunit](https://github.com/xunit/abstractions.xunit/blob/master/src/xunit.abstractions/Frameworks/ITestFrameworkOptions.cs) we have similar collection but it using a string-key instead, liking this model better.

Reply:
Why?

Reply:
I don't need to remember the magic strings for the key. I also have a feeling thst it use less allocations but nothing that I have run profiler on.

Reply:
No, that one makes sense :) The Set thing I don't understand. It seems like an extension shouldn't need to call Set twice, so Add seems better to me.

Reply:
I think they prefere to be able to set an item multiple time (updating values) without first checking if the key already exists and in that case remove before the add can occur. If using Add and we should simulate the dictionary we should also throw exception on Get if key not exists.

The xunit also use GetValue/SetValue.

@jbogard any specific taste for Get/Add/Remove vs. Get/Set?

Reply:
Is there a data structure that already matches the behavior we're looking for?

Reply:
It's a dictionary as implementation, but I think the idea is to expose a restricted interface. I, for one, don't see a need for anything besides GetOrAdd. This would ensure a single feature per extension and it would be easy to use.

Reply:
I should say that sometimes it need to replace the data for a specific extension, example
```
cfg.CreateMap<OrderDomain, OrderEf>()
    .EqualityComparison((oo, dto) => oo.Id == dto.Id)
    .ReverseMap()
    .EqualityComparison((dto, oo) => dto.AnotherId == oo.AnotherId);
```
where the `ReverseMap()` will add an reverse version of the `.EqualityComparison((oo, dto) => oo.Id == dto.Id)` and then it will be overwritten by `.EqualityComparison((dto, oo) => dto.AnotherId == oo.AnotherId)`. It also depend of if the item that is stored in the feature collection is an immutable object or not. If the object is immutable we need to replace the object, with a new object, otherwise it can probably be updated but we in that case always need to use logic to check if the object exists as a feature. If we have a `Set` method only then we always overwrite the item with the new item and don't care if that was existing before adding the new object. That means that only Get/Add not is enough and we need to have Get/Add/Remove in that case.

Reply:
You could replace the object as you say, or you can only have GetOrAdd and update an existing object.

Reply:
I changed the name to `Add` instead and it will have the same logic as the indexed property `this[type] = value` so it replaces the value if it is already existing.

Reply:
That seems misleading.

Reply:
Depend on how the feature class is build up I in that case prefer `AddOrUpdate` instead of `GetOrAdd` to not always need to put in an delegate for add even if I only want to check if the feature exists.

When you write code that is aware or the feature and will check if that exists before adding need to add that if-logic in the calling code.